baseplate/queue_consumer.py

```diff
@@ -0,0 +1,127 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import contextlib
+from threading import Thread
+
+from baseplate._compat import queue
+from kombu import Queue
+from kombu.mixins import ConsumerMixin
+
+
+def consume(baseplate, exchange, connection, queue_name, routing_keys, handler):
+    """Create a long-running process to consume messages from a queue.
+
+    A queue with name ``queue_name`` is created and bound to the
+    ``routing_keys`` so messages published to the ``routing_keys`` are routed
+    to the queue.
+
+    Next, the process registers a consumer that receives messages from
+    the queue and feeds them to the ``handler``.
+
+    The ``handler`` function must take 3 arguments:
+
+    * ``context``: a baseplate context
+    * ``message_body``: the text body of the message
+    * ``message``: :py:class:`kombu.message.Message`
+
+    The consumer will automatically ack each message after the handler method
+    exits. If there is an error in processing and the message must be retried
+    the handler should raise an exception to crash the process. This will
+    prevent the ack and the message will be requeued at the head of the queue.
+
+    :param baseplate.core.Baseplate baseplate: A baseplate instance for the
+        service.
+    :param kombu.Exchange exchange:
+    :param kombu.connection.Connection connection:
+    :param str queue_name: The name of the queue.
+    :param list routing_keys: List of routing keys.
+    :param handler: The handler method.
+
+    """
+    queues = []
+    for routing_key in routing_keys:
+        queue = Queue(
+            name=queue_name,
+            exchange=exchange,
+            routing_key=routing_key,
+        )
+        queues.append(queue)
+
+    print("registering %s as a handler for %r" % (handler.__name__, queues))
+    kombu_consumer = KombuConsumer(baseplate, connection, exchange, queues)
+    kombu_consumer.start()
+
+    print("waiting for messages")
+    while True:
+        with kombu_consumer.get_message() as (context, message, span):
+            handler(context, message.body, message)
+            message.ack()
+
+
+class _ConsumerWorker(ConsumerMixin):
+    def __init__(self, connection, queues, work_queue):
+        self.connection = connection
+        self.queues = queues
+        self.work_queue = work_queue
+
+    def get_consumers(self, Consumer, channel):
+        return [Consumer(
+            queues=self.queues,
+            on_message=self.on_message,
+        )]
+
+    def on_message(self, message):
+        self.work_queue.put(message)
+
+
+class ConsumerContext(object):
+    pass
+
+
+class KombuConsumer(object):
```

Comment:
I think the direct interface could be useful to expose as well for more complex users. Can this be exposed and documented? (Modulo potential changes for splitting start/get_)